optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java

```diff
@@ -116,20 +127,26 @@ public Solution_ call() {
 
     @Override
     public void terminateEarly() {
-        boolean cancelled = future.cancel(false);
-        if (cancelled) {
-            solverStatus = SolverStatus.NOT_SOLVING;
-        } else {
-            // The solver is either actively solving or has already terminated
-            solver.terminateEarly();
-            try {
-                // Don't return until bestSolutionConsumer won't be called any more
-                terminatedLatch.await();
-            } catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
-                logger.warn("The terminateEarly() call is interrupted.", e);
+        future.cancel(false);
+        synchronized (this) {
+            if (solverStatus == SolverStatus.SOLVING_SCHEDULED) {
+                solvingTerminated();
+            } else if (solverStatus == SolverStatus.SOLVING_ACTIVE) {
+                // Indirectly triggers solvingTerminated()
+                solver.terminateEarly();
+            } else if (solverStatus == SolverStatus.NOT_SOLVING) {
+                // Do nothing, solvingTerminated() already called
+            } else {
+                throw new IllegalStateException("Unsupported solverStatus (" + solverStatus + ").");
             }
```

Comment:
This should be replacable by `switch (solverStatus.get())`, as all the actual state mutation happens in `solverTerminated()`/`solverTerminateEarly()` as I've suggested above.